### DISCUSSION TITLE
Updating field in XFA PDF

### DISCUSSION THREAD

**Discussion by jheasleycrea** [py-pdf/pypdf#2809] [Q&A] [answered]
I am attempting to update some fields in an XFA enabled pdf. So far the closest I've come to this is finding /XFA and accessing the dataset portion via `xfa[7].get_object().get_data()`

This returns the XML that corresponds to all the field values. From there I update the xml with my desired values and then attempt to set it via `xfa[7].get_object().set_data(write_xml)`.

This throws an Exception: `pypdf.errors.PdfReadError: Streams encoded with different filter from only FlateDecode is not supported`.

I am extremely new to PDF manipulation and wouldn't be surprised if I'm going entirely in the wrong direction, but does anyone have any advice on how to achieve this?

Edit:

```python
import pypdf
def findInDict(needle, haystack):
    for key in haystack.keys():
        try:
            value=haystack[key]
        except:
            continue
        if key==needle:
            return value
        if isinstance(value,dict):            
            x=findInDict(needle,value)            
            if x is not None:
                return x
pdfobject=open('Test.pdf','rb')
pdf=pypdf.PdfReader(pdfobject)
xfa=findInDict('/XFA',pdf.resolved_objects)

xml=xfa[7].get_object().get_data()
write_xml = xml.replace(b'101101101</ce:PartyIdentificationNumberText\n>', b'202202202</ce:PartyIdentificationNumberText\n>')
xfa[7].get_object().set_data(write_xml)


writer = pypdf.PdfWriter(clone_from=pdf)
outputstream = open('write_file.pdf', 'wb')
writer.write(outputstream)

outputstream.close()
pdf.close()
```

**Comment by stefan6419846** [comment]
Please provide some complete code example and an example PDF for further analysis.

The error message indicates that the object you are trying to replace does not use the *FlateDecode* filter. *pypdf* currently does not know how to encode data with other filters, thus your approach does not work here at the moment.

**Reply by jheasleycrea** [reply]
[write_file.pdf](https://github.com/user-attachments/files/16729350/write_file.pdf)
Here's the file. 
The code so far is edited above

**Reply by pubpub-zz** [reply]
@jheasleycrea 
can you send also the original file.thx

**Reply by jheasleycrea** [reply]
Here's the original
[Test.pdf](https://github.com/user-attachments/files/16731111/Test.pdf)

**Reply by jheasleycrea** [reply]
To be clear. I originally wiped write_file.pdf before attaching it here. Currently if I don't try and update the data, the PDFWriter successfully spits out a copy of Test.pdf that is identical as far as I can tell

**Comment by jheasleycrea** [comment, marked as answer]
Figured out the problem. The issue from the fact that the /Filter here was an `ArrayObject` of `["/FlateDecode"]`. in `EncodedStreamObject.set_data` the line to check Decode status `if self.get(SA.FILTER, "") == FT.FLATE_DECODE:` returns false.
I was able to work around this by setting /Filter to `_base.NameObject("/FlateDecode")` 
From there the rest of my code ran and the field was updated as expected.

**Reply by pubpub-zz** [reply]
You did good! sorry for not answering earlier.
for the value you should use `pypdf.generic.NameObject("/FlateDecode")` I'm wondering if you should not also set the field value

**Reply by ljbergmann** [reply]
Do you have an example for this? I just looked at the documentation and i can't find anything relating to this topic and i'm a bit lost here.

**Reply by pubpub-zz** [reply]
have a look here:
https://pypdf.readthedocs.io/en/stable/modules/generic.html

by the way, I've produced PR #2821 that will make set_data compatible with your case

**Reply by ljbergmann** [reply]
Thank you very much and indeed it is working ;) combined with #2811 I'm able to fill out the pdf. Now only the error message problem is left. I'm trying to identify why acrobat is throwing the error message and maybe i can contribute to the PR / project in finding a solution for this.

**Reply by pubpub-zz** [reply]
Which error message ?

**Reply by ljbergmann** [reply]
the one i mentioned in #2811

**Comment by ericxiao251** [comment]
Is there a full code example? I tried updating a field in the form and datasets xml but nothing happened.

**Reply by stefan6419846** [reply]
The original code is for a very specific case, only replacing the plain `101101101</ce:PartyIdentificationNumberText\n>` value in the XML data stream. Thus the code will do nothing unless you happen to run it on exactly the same PDF file.
